src/protocols/output_power_management.rs

```diff
@@ -0,0 +1,238 @@
+    where
+        D: GlobalDispatch<ZwlrOutputPowerManagerV1, OutputPowerManagerGlobalData>,
+        D: Dispatch<ZwlrOutputPowerManagerV1, ()>,
+        D: Dispatch<ZwlrOutputPowerV1, OutputPowerState>,
+        D: OutputPowerManagementHandler,
+        D: 'static,
+        F: for<'c> Fn(&'c Client) -> bool + Send + Sync + 'static,
+    {
+        let global_data = OutputPowerManagerGlobalData {
+            filter: Box::new(filter),
+        };
+        display.create_global::<D, ZwlrOutputPowerManagerV1, _>(VERSION, global_data);
+        Self {
+            output_powers: HashMap::new(),
+        }
+    }
+
+    pub fn output_removed(&mut self, output: &Output) {
+        if let Some(output_power) = self.output_powers.remove(output) {
+            output_power.failed();
+        }
+    }
+
+    pub fn output_power_mode_changed(&mut self, output: &Output, mode: Mode) {
+        if let Some(output_power) = self.output_powers.get(output) {
+            output_power.mode(mode.into());
+        }
+    }
+}
+
+impl<D> GlobalDispatch<ZwlrOutputPowerManagerV1, OutputPowerManagerGlobalData, D>
+    for OutputPowerManagementManagerState
+where
+    D: GlobalDispatch<ZwlrOutputPowerManagerV1, OutputPowerManagerGlobalData>,
+    D: Dispatch<ZwlrOutputPowerManagerV1, ()>,
+    D: 'static,
+{
+    fn bind(
+        _state: &mut D,
+        _handle: &DisplayHandle,
+        _client: &Client,
+        manager: New<ZwlrOutputPowerManagerV1>,
+        _global_data: &OutputPowerManagerGlobalData,
+        data_init: &mut DataInit<'_, D>,
+    ) {
+        data_init.init(manager, ());
+    }
+
+    fn can_view(client: Client, global_data: &OutputPowerManagerGlobalData) -> bool {
+        (global_data.filter)(&client)
+    }
+}
+
+impl<D> Dispatch<ZwlrOutputPowerManagerV1, (), D> for OutputPowerManagementManagerState
+where
+    D: Dispatch<ZwlrOutputPowerManagerV1, ()>,
+    D: Dispatch<ZwlrOutputPowerV1, OutputPowerState>,
+    D: OutputPowerManagementHandler,
+    D: 'static,
+{
+    fn request(
+        state: &mut D,
+        _client: &Client,
+        _resource: &ZwlrOutputPowerManagerV1,
+        request: <ZwlrOutputPowerManagerV1 as Resource>::Request,
+        _data: &(),
+        _dhandle: &DisplayHandle,
+        data_init: &mut DataInit<'_, D>,
+    ) {
+        match request {
+            zwlr_output_power_manager_v1::Request::GetOutputPower { id, output } => {
+                if let Some(output) = Output::from_resource(&output) {
+                    // We borrow state in the middle.
+                    #[allow(clippy::map_entry)]
+                    if !state
+                        .output_power_manager_state()
+                        .output_powers
+                        .contains_key(&output)
+                    {
+                        let zwlr_output_power = data_init.init(id, OutputPowerState {});
+                        zwlr_output_power.mode(state.get_output_power_mode(&output).into());
```

Comment:
So yeah as I wrote in the other comment, the way this should work is `OutputPowerManagerState` should itself store the current power mode (to its knowledge) and send that straight to clients. Then there should be a `refresh()` call that tells it the current up-to-date power mode from niri. If it changes, it will send that update to clients.